### ISSUE TITLE
White bar appears when using external keyboard on iPad

### ISSUE THREAD

**Issue by noisymime** [microsoft/vscode#149048]
Does this issue occur when all extensions are disabled?: Yes

* Device: iPad Pro
* code-server: v4.4.0
* VSCode: 1.66.2

Steps to Reproduce:
Using an iPad with external keyboard, selecting an area that allows keyboard input causes a white bar to appear at the bottom of the screen:

![90480EB9-A3EF-4E35-BF99-EB8863513749](https://user-images.githubusercontent.com/138502/167321887-9b17cbec-4899-4ab7-9780-aa667aac00ab.png)

This is similar to #122390, though I am running a version after that fix has been added. 
This occurs in both PWA mode and regular Safari. Problem does not appear when using the in-built keyboard

The bar disappears when the keyboard is not active/usable.

**Comment by TylerLeonhardt** [comment]
Yes this seems to have occurred after Apple went from a full bottom bar to this hovering bar at the bottom.

I think we probably want to stretch VS Code too the very bottom (thus having some of it behind the hovering iPad bar). I'm a bit concerned that then the floating bar will be in the way... but that seems to be the canonical approach on iPad.

**Comment by cschoenecker** [comment]
Same bug, very annoying

**Comment by TylerLeonhardt** [comment]
I'm marking this as `help wanted` in case someone wants to investigate this. I have a feeling some magic will need to be added around here similar to what was there before:
https://github.com/microsoft/vscode/commit/dbed1fbe954fdd34b05deeb273198d4f539a6173

Not sure how other websites handle this.

**Comment by cschoenecker** [comment]
> You can go to `Settings > General> Keyboards` and disable `shortcuts`. with this workaround you can get rid of that annoying floating button, and vscode works perfectly.

Yes thanks, I discover this workaround but we have to disable “Shortcut” and the second above option on iPadOS 15 to hide the keyboard shortcut :)

**Comment by vultuk** [comment]
> You can go to `Settings > General> Keyboards` and disable `shortcuts`. with this workaround you can get rid of that annoying floating button, and vscode works perfectly.

Unfortunately this also means that scrolling with a trackpad or mouse no longer works.

Interestingly I just had a check and the shorcut issue does not happen on Github Codespaces, the shortcut bar is there, but there's no white space.

**Comment by cthos** [comment]
I noticed this on Codespaces too, but it does appear if you switch tabs and return to the Codespace (though it does disappear if you subsequently reload the window).

**Comment by icharle** [comment]
> > You can go to `Settings > General> Keyboards` and disable `shortcuts`. with this workaround you can get rid of that annoying floating button, and vscode works perfectly.
> 
> Unfortunately this also means that scrolling with a trackpad or mouse no longer works.
> 
> Interestingly I just had a check and the shorcut issue does not happen on Github Codespaces, the shortcut bar is there, but there's no white space.

I had the same problem, disable shortcuts would cause that scrolling no works. it isn't look like a perfect solution.

**Comment by TylerLeonhardt** [comment]
I opened an issue on WebKit as this appears to be on them to fix since visualViewport isn't what I would have expected:
https://bugs.webkit.org/show_bug.cgi?id=247410

**Comment by fosron** [comment]
+1 on scrolling not working on iPad Air with external keyboard

**Comment by filiptronicek** [comment]
Hey there @TylerLeonhardt 👋! Has there been any updates on your side with this issue? It still seems to happen in both Gitpod and https://github.com/coder/code-server, but I could not reproduce in Codespaces or https://github.dev.

**Comment by TylerLeonhardt** [comment]
I was always waiting on a WebKit fix... doesn't look like there's any traction in that issue. I'll see if there's anything we can on our end to push for a fix...

but you bring up a good point... how did they fix that? 😆 

Let me ask them and see if I can get to the bottom of that.

**Comment by TylerLeonhardt** [comment]
Looked at this with @osortega and I think I know why this is happening. Looking here:

https://github.com/microsoft/vscode/blob/5e805b79fcb6ba4c2d23712967df89a089da575b/src/vs/base/browser/dom.ts#L291-L299

this is called with the workbench's parent [here](https://github.com/microsoft/vscode/blob/5e805b79fcb6ba4c2d23712967df89a089da575b/src/vs/workbench/browser/layout.ts#L1380).

In vscode.dev, the workbench's parent is `body`, so it goes to that second `if`:

![Image](https://user-images.githubusercontent.com/2644648/224178641-efe25df5-d2e4-4a6d-9857-7e4f734fbc7f.png)

But in github.dev, the workbench's parent is this other `div` so it goes to the first `if`:

![Image](https://user-images.githubusercontent.com/2644648/224178711-f9f445fb-a791-4e07-a237-06d11676e5aa.png)

Well that changes the behavior of `getClientArea` using `element.clientHeight` instead of `window.visualViewport.height`.

The `clientHeight` of that random `div` is the correct value. Walking up the dom, we see:
* `<div id=​"serverless-workbench" style=​"height:​ 100%;​">​` so `height: 100%`
* `<div class=​"vsonline-serverless-workbench">​` which has an interesting:
```css
.vsonline-serverless-workbench {
  position: absolute;
  left: 0;
  top: 0;
  right: 0;
  bottom: 0;
  z-index: 3;
}
```
* `<div id=​"js-vscode-workbench">​` that has nothing interesting
* `<body class="mac web">` has the same stuff as the body of vscode.dev
* `<html>` nothing interesting

So that's what is happening and why... now the question is how should it be fixed.

**Comment by TylerLeonhardt** [comment]
Ok so github.dev's solution works, but it has the following two problems:
1. status bar is covered by home indicator:
![image](https://user-images.githubusercontent.com/2644648/224194744-2bf02a89-8d0f-49e5-a846-3e68db4d21b4.png)
2. on screen keyboard doesn't cause a resize:
![image](https://user-images.githubusercontent.com/2644648/224194783-cb613d30-e2bb-42e8-9833-f9eb53f0f76e.png)
    * So you might not know where you are typing
    * and you can't see the status bar or terminal

Because of these points, I actually think the vscode.dev experience is better despite the big white bar. So I think that a solution for the WebKit bug is absolutely the right way forward to fix this bug. Only that, will achieve the best result of:
* height will be right above the height of the home indicator so it doesn't cover the status bar
* height will dynamically change if the on-screen keyboard shows up

vscode.dev could probably do a better job of honoring the theme color... so at least instead of white, it's the same color as the editor or something...

**Comment by filiptronicek** [comment]
Many thanks to you both for digging into this, @TylerLeonhardt, your findings have been incredibly helpful. 

Agreed the github.dev's solution is not ideal either. Let's hope Apple moves with the WebKit bug soon!

**Comment by kevin-king** [comment]
> Because of these points, I actually think the vscode.dev experience is better despite the big white bar.

I disagree, the permanent white bar is significantly worse UX than a line over the status bar or the virtual keyboard covering. One reason is that the white bar affects users at all times whereas using the on-screen keyboard is less impactful since users almost certainly have a hardware keyboard plugged in when using vscode and can still scroll the text above the virtual keyboard.

Screen real estate on iPad devices is additionally hindered by this bug https://github.com/coder/code-server/issues/3203 (upstream issue) where you lose text selection capabilities in full-screen mode and thus can't use the full view width-wise either. 

From my testing, the experience is a lot better if we use innerHeight on iPad:

```ts
export function getClientArea(element: HTMLElement): Dimension {

        if (platform.isIOS && window.innerWidth && window.innerHeight) {
            return new Dimension(window.innerWidth, window.innerHeight);
        }

	// Try with DOM clientWidth / clientHeight
	if (element !== document.body) {
		return new Dimension(element.clientWidth, element.clientHeight);
	}

	// Try innerWidth / innerHeight
	if (window.innerWidth && window.innerHeight) {
		return new Dimension(window.innerWidth, window.innerHeight);
	}

	// Try with document.body.clientWidth / document.body.clientHeight
	if (document.body && document.body.clientWidth && document.body.clientHeight) {
		return new Dimension(document.body.clientWidth, document.body.clientHeight);
	}

	// Try with document.documentElement.clientWidth / document.documentElement.clientHeight
	if (document.documentElement && document.documentElement.clientWidth && document.documentElement.clientHeight) {
		return new Dimension(document.documentElement.clientWidth, document.documentElement.clientHeight);
	}

	throw new Error('Unable to figure out browser width and height');
}
```

**Comment by TUCAOEVER** [comment]
这是来自QQ邮箱的假期自动回复邮件。
 
您好，我最近正在休假中，无法亲自回复您的邮件。我将在假期结束后，尽快给您回复。

**Comment by kevin-king** [comment]
Here was a previous attempt made to solve this issue and the patch which caused the regression: https://github.com/microsoft/vscode/issues/122390

**Comment by kjswanson** [comment]
Long time lurker first time poster.

I have side stepped the issue by using code-server v4.12.0. I start the code-server pwa on the ipad and then drag it to the external screen. This has been working for a couple of weeks now.

Hope this information gives someone in the “know” a good lead on the issue.

**Comment by TUCAOEVER** [comment]
这是来自QQ邮箱的假期自动回复邮件。
 
您好，我最近正在休假中，无法亲自回复您的邮件。我将在假期结束后，尽快给您回复。

**Comment by halfama** [comment]
Still no solution for that ?

**Comment by TUCAOEVER** [comment]
这是来自QQ邮箱的假期自动回复邮件。
 
您好，我最近正在休假中，无法亲自回复您的邮件。我将在假期结束后，尽快给您回复。

**Comment by TylerLeonhardt** [comment]
No progress here as there has been no progress by WebKit :( https://bugs.webkit.org/show_bug.cgi?id=247410 feel free to add fuel to the fire... maybe more comments will give more eyes.

**Comment by lukaszciastko** [comment]
I’m curious… Could you at least make it configurable? A lot has changed since this bug was reported. I now frequently use AI agents and often leave my laptop at home, carrying only my iPad to remotely check on work via VS Code tunnels. This particular bug is quite frustrating.
